File: example/customersdotnet/Service.cs
using System.Threading.Tasks;

namespace Customers
{
    public class Service
    {
        private Outbound outbound;

        public Service(Outbound outbound)
        {
            this.outbound = outbound;
        }

        public async Task<Customer> CreateCustomer(Customer customer)
        {
            await outbound.SaveCustomer(customer);
            await outbound.CustomerCreated(customer);

            return customer;
        }

        public Task<Customer> GetCustomer(ulong id)
        {
            return outbound.FetchCustomer(id);
        }
    }
}

File: example/customers_wasm/assembly/service.ts
import { consoleLog as log } from "@wapc/as-guest";
import {
  Service,
  Customer,
  CustomerQuery,
  CustomerPage,
  Repository,
  Publisher,
} from "./interfaces";

export class ServiceImpl implements Service {
  private repository: Repository;
  private publisher: Publisher;

  constructor(repository: Repository, publisher: Publisher) {
    this.repository = repository;
    this.publisher = publisher;
  }

  createCustomer(customer: Customer): Customer {
    log("createCustomer called");
    this.repository.saveCustomer(customer);
    this.publisher.customerCreated(customer);

    return customer;
  }

  getCustomer(id: u64): Customer {
    log("getCustomer called");
    return this.repository.fetchCustomer(id);
  }

  listCustomers(query: CustomerQuery): CustomerPage {
    log("listCustomers called");
    return CustomerPage.newBuilder()
      .withOffset(query.offset)
      .withLimit(query.limit)
      .build();
  }
}
